### ISSUE TITLE
Infer numpy dtypes as numba.types.scalars.xxxx

### ISSUE THREAD

**Issue by sjperkins** [numba/numba#3503]
## Feature request

It would be useful for the numba type inference system to infer arguments containing numpy types as numba.types.scalars.xxxx.

```python
from __future__ import print_function

import numba
import numpy as np

NoneTypes = (numba.types.misc.NoneType, numba.types.misc.Omitted)


@numba.generated_jit(nopython=True, nogil=True)
def fn(a, b, dtype=None):

    print("Initial dtype", dtype)

    # In case of None, infer type from present input args
    if isinstance(dtype, NoneTypes):
        # Have to produce a numpy dtype here from the numba type names
        out_dtype = np.result_type(*(np.dtype(ary.dtype.name)
                                   for ary in (a, b)
                                   if not isinstance(ary, NoneTypes)))
    else:
        # We should have a proper type here, but numba infers the
        # type of np.int32 as a pyobject, rather than
        # types.misc.npytypes.DType?
        out_dtype = numba.numpy_support.as_dtype(dtype)

    print("final dtype", out_dtype)

    def _fn(a, b, dtype=None):
        return np.ones(a.shape, out_dtype)

    return _fn


a = np.zeros(10)
b = np.zeros(10)

# Succeeds because dtype=None
print("fn(a, b)", fn(a, b))
# Fails because dtype=np.int32 is inferred as a pyobject
print("fn(a, b, dtype=np.int32)", fn(a, b, dtype=np.int32))
```

I'd imagine this is a fairly simple change, but I find the type system a bit opaque. Could a maintainer suggest where the change might be made?

**Comment by sklam** [comment]
We are missing a `typeof` implementation for numpy types.  It should be registered into `numba/typing/typeof.py` (i.e. https://github.com/numba/numba/blob/e74336755bbba68d800985bd470fa35b9d27d3d3/numba/typing/typeof.py#L196-L199)

It's also missing unboxing logic in `numba/targets/boxing.py`.

**Comment by sjperkins** [comment]
Closed by #3516
